### DISCUSSION TITLE
Serialization of UUID

### DISCUSSION THREAD

**Discussion by bonzobuufo** [fastapi/fastapi#3239] [Questions] [answered]
Hi,

Thank-you for a great library!  It really takes the grind out of writing an API!

I'm working on a project that is using FastAPI v0.62.0 to provide an API web service, and using Pydantic v1.7.3 also to help with the payload data typing.  The project is written in Python 3.9.  I've had no problems using standard library UUID's as part of the URL path, and also when using them as part the response defined in the Pydantic User Model, where that is defined in the "response_model" of the path decorator.  However, one of my API services needs to have alternative User Models in it's possible responses - I've done this successfully using the "responses" argument to the path decorator... until I wanted to define a UUID in the payload.  It seems that where you define a response_model User Model that happens to contain UUID's - the library can automatically serialise the UUID in the User Model when you return the UM as the return value.   However - if I don't have UUID defined in the UM for the response_model, and instead try to use it in a UM that is specifed in the "responses" list - then when using JSONResponse to return a result - the "someum.dict()" method issues an error:

`TypeError: Object of type UUID is not JSON serializable`

It seems that JSON Serialisation works for the main UM, but not when you try to use JSONResponse to serialise it...

Is this a bug?

Attached 
[testuuid.zip](https://github.com/tiangolo/fastapi/files/6491714/testuuid.zip)
is some source code demonstrating the "bug" (method "uuidaltfail") - the error appear in the logs.  The last two functions are workarounds to overcome the problem.

**Comment by negadive** [comment]
From the [doc](https://fastapi.tiangolo.com/advanced/response-directly/#return-a-response)
>And when you return a Response, FastAPI will pass it directly.
It won't do any data conversion with Pydantic models, it won't convert the contents to any type, etc.
This gives you a lot of flexibility. You can return any data type, override any data declaration or validation, etc

JSONResponse itself is a sub-class of Response. So when you pass `pl.dict()` to JSONResponse:
- fastapi didnt do any convertion, uses `json.dumps` instead to serialize
- the passed value is still instance of `UUID` which cannot serialized by `json.dumps`

**Reply by tiangolo** [reply]
Thanks @negadive! ☕ 

@bonzobuufo you probably need to read the docs about the `jsonable_encoder`: https://fastapi.tiangolo.com/tutorial/encoder/

**Comment by tiangolo** [comment, marked as answer]
@bonzobuufo you probably need to read the docs about the jsonable_encoder: [fastapi.tiangolo.com/tutorial/encoder](https://fastapi.tiangolo.com/tutorial/encoder/)

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
